### PR TITLE
feat: add round function

### PR DESCRIPTION
**Pull request by vibhatha**
This PR adds `round` function.

### CODE REVIEW

**Comment by jvanstraten** [comment]
Where did this come from? Based on #307 I thought we weren't going to do rounding to multiple. Rounding to some number of digits is exactly the same thing, just restricted to powers of 10.

That rounding argument is also really weird, there's no such thing as an optional value argument and using an integer for this is really weird. If we *are* going to do a round-to-multiple I feel like we should just reuse the existing rounding mode option to determine how to round.

Based on #307 I would expect simply `round(f32) -> f32` and `trunc(f32) -> f32` (and likewise for `f64`).

**Comment by vibhatha** [comment]
> Where did this come from? Based on #307 I thought we weren't going to do rounding to multiple. Rounding to some number of digits is exactly the same thing, just restricted to powers of 10.
> 
> That rounding argument is also really weird, there's no such thing as an optional value argument and using an integer for this is really weird. If we _are_ going to do a round-to-multiple I feel like we should just reuse the existing rounding mode option to determine how to round.
> 
> Based on #307 I would expect simply `round(f32) -> f32` and `trunc(f32) -> f32` (and likewise for `f64`).

Referring to: https://github.com/substrait-io/substrait/pull/230/files#diff-b6317d19776a0381379a5980e1be0ffcde22082a034b891cb7a072fdbab32327

**Comment by vibhatha** [comment]
@jvanstraten I was referring to the SQL definition [here](https://docs.microsoft.com/en-us/sql/t-sql/functions/round-transact-sql?view=sql-server-ver16).

I see your point on `function` argument. In Duckdb there is no such parameter. May be I will remove that option and also add the rounding options here too. 

WDYT? cc @ianmcook

**Comment by ianmcook** [comment]
IIUC, I think we can just implement an rounding mode option, and make `TRUNCATE` one of the available modes, alongside the others like `DOWN`, `UP`, `TOWARDS_ZERO`, etc. Does that sound right to you @jvanstraten ?

**Comment by jvanstraten** [comment]
I figured we could just reuse the floating point rounding mode. I also generalized it to integers for quantile earlier. Wouldn't `TRUNCATE` and `TOWARDS_ZERO` be the same thing, anyway?

Note that dedicated ceil and floor functions already exist. I guess we should get rid of those if we go the option route.

**Comment by ianmcook** [comment]
Ah, yes, I think that all makes sense.

Since a lot of thought went into ensuring that [Acero's `RoundMode`](https://arrow.apache.org/docs/cpp/api/compute.html#_CPPv49RoundMode) can represent the behavior of any other mainstream system's round function, I'll list out the options it provides and map them onto the options used in the Substrait functions that take a `rounding` argument. To see an example showing the behavior of these different Acero `RoundMode`, look at [this test](https://github.com/apache/arrow/blob/dd0988b49cb6726cf915bb9f53d7320e3a97b00b/python/pyarrow/tests/test_compute.py#L1509-L1521).

- `DOWN` - corresponds to Substrait's `FLOOR`
- `UP` - corresponds to Substrait's `CEILING`
- `TOWARDS_ZERO` - corresponds to Substrait's `TRUNCATE`
- `TOWARDS_INFINITY` - no corresponding option in Substrait
- `HALF_DOWN` -  no corresponding option in Substrait
- `HALF_UP` - no corresponding option in Substrait
- `HALF_TOWARDS_ZERO` - no corresponding option in Substrait
- `HALF_TOWARDS_INFINITY` - corresponds to Substrait's `TIE_AWAY_FROM_ZERO`
- `HALF_TO_EVEN` - corresponds to Substrait's `TIE_TO_EVEN` 
- `HALF_TO_ODD` - no corresponding option in Substrait

So this raises some questions:
- Should we add a Substrait option `AWAY_FROM_ZERO`? This corresponds to Acero's `TOWARDS_INFINITY `.
- Should we add a Substrait option `TIE_TOWARDS_NEGATIVE_INFINITY `? This corresponds to Acero's `HALF_DOWN`.
- Should we add a Substrait option `TIE_TOWARDS_POSITIVE_INFINITY `? This corresponds to Acero's `HALF_UP`.
- Should we add a Substrait option `TIE_TOWARDS_ZERO`? This corresponds to Acero's `HALF_TOWARDS_ZERO`.
- Should we add a Substrait option `TIE_TO_ODD`? This corresponds to Acero's `HALF_TO_ODD`.

**Comment by jvanstraten** [comment]
Hm. I guess that makes sense for round, and adding more variants is cheap. I don't think it makes sense to do it for all floating point functions though, since IEEE is pretty explicit about what the possible rounding modes are.

**Comment by vibhatha** [comment]
cc @jvanstraten I added the round options. Does it reflect what you expected?

**Comment by jvanstraten** [comment]
No, because you didn't add the rounding modes @ianmcook was talking about, and the supposedly-optional `function` argument still exists.

**Comment by vibhatha** [comment]
@jvanstraten I added other options. Should we create documentation for this too? In this PR itself or a separate PR?

**Comment by jvanstraten** [comment]
It should just be in the description of the function in this case; this option argument is basically specific to `round`. The rounding mode should probably be a required argument, too, since it makes little sense for a producer to allow a consumer to round however it wants for a function that is specifically about rounding (unlike the usual floating point rounding mode).

**Comment by vibhatha** [comment]
@jvanstraten I had the doubt though. I will update it.

**Review by westonpace — CHANGES_REQUESTED** [review]
@vibhatha when you get a moment can you update this to the new format for optional arguments?

**Comment by vibhatha** [comment]
> @vibhatha when you get a moment can you update this to the new format for optional arguments?

sure @westonpace, thanks for the ping. I will update it.

File: extensions/functions_rounding.yaml [hunk 1]
```diff
@@ -27,3 +27,68 @@ scalar_functions:
...
+  -
+    name: "round"
+    description: >
+      Rounding the value `x` to s decimal places.
+
+    impls:
+      - args:
+          - name: rounding
+            options: [ TIE_TO_EVEN, TIE_AWAY_FROM_ZERO, TRUNCATE, CEILING, FLOOR, AWAY_FROM_ZERO,
+              TIE_DOWN, TIE_UP, TIE_TOWARDS_ZERO, TIE_TOWARDS_ZERO, TIE_TO_ODD ]
```
Comment (EpsilonPrime) [inline extensions/functions_rounding.yaml:39]: TIE_TOWARDS_ZERO is listed twice on this line.

File: extensions/functions_rounding.yaml [hunk 2]
```diff
@@ -27,3 +27,68 @@ scalar_functions:
...
+              This is an optional argument `function`, a numeric
+              value determines whether to perform round or truncate.
+              When the `function` value is set to 0, the function
+              performs rounding to `s` decimal places, otherwise it
+              performs truncate. The default value is 0.
+        nullability: DECLARED_OUTPUT
+        return: fp32?
+      - args:
+          - name: rounding
+            options: [ TIE_TO_EVEN, TIE_AWAY_FROM_ZERO, TRUNCATE, CEILING, FLOOR ]
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply extensions/functions_rounding.yaml]: no that is a typo.

File: extensions/functions_rounding.yaml (hunk 1 above)
Comment (vibhatha) [reply extensions/functions_rounding.yaml]: Thanks for noticing this, it is a typo.

File: extensions/functions_rounding.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (vibhatha) [reply extensions/functions_rounding.yaml]: @EpsilonPrime  I updated the PR. Please take a look.

File: extensions/functions_rounding.yaml [hunk 3]
```diff
@@ -27,3 +27,68 @@ scalar_functions:
...
+              performed to the left side of the decimal point
+              as specified by `s`.
+          - value: i32
+            name: "function"
+            description: >
+              This is an optional argument `function`, a numeric
+              value determines whether to perform round or truncate.
+              When the `function` value is set to 0, the function
+              performs rounding to `s` decimal places, otherwise it
+              performs truncate. The default value is 0.
```
Comment (westonpace) [inline extensions/functions_rounding.yaml:58]: Why isn't this an enum?

Comment (vibhatha) [reply extensions/functions_rounding.yaml:58]: @westonpace I think roughly I get your point. But could you please give me an example here, I don't think my understanding is very precise.

**Review by westonpace — CHANGES_REQUESTED** [review]
Sorry, my previous comments were terse and I will explain in more detail:

There are currently three value arguments, `x`, `s`, and `function`.  No engine I am aware of implements a 3-arg round function.  I have a few concerns:

1. `function` is redundant and should be a bool or an enum instead of an `i32`.  It currently is used to determine whether we are rounding or truncating.  However, "truncate" is (as far as I know) equivalent to "round" with `rounding = FLOOR`.

So I think we can get rid of `function`.  If I am misunderstanding it, and we choose to keep it, then it should not be `i32` but something like:

```
value: bool
name: "should_truncate"
description: >
  If true perform truncation.  If false then round to `s` decimal places.  The default value is false.
```

2. Should `s` be an option or a value argument?

In other words, is `round` a unary function with an option telling it how much to round to?  Or is `round` a binary function?

My initial thought was "unary function" since that is how it is implemented in Acero.  However, looking further, it appears that other engines have implemented this as a binary function.  So I think binary function is the correct answer and this comment is no longer valid.

Let me know how you'd like to proceed on question 1.

**Comment by westonpace** [comment]
Oops, one minor clarification. I said `function != 0` is equivalent to `rounding = FLOOR` but I meant `function != 0` is equivalent to `rounding = TRUNCATE`.

**Comment by vibhatha** [comment]
> Sorry, my previous comments were terse and I will explain in more detail:
> 
> There are currently three value arguments, `x`, `s`, and `function`. No engine I am aware of implements a 3-arg round function. I have a few concerns:
> 
> 1. `function` is redundant and should be a bool or an enum instead of an `i32`.  It currently is used to determine whether we are rounding or truncating.  However, "truncate" is (as far as I know) equivalent to "round" with `rounding = FLOOR`.
> 
> **So I think we can get rid of `function`.** If I am misunderstanding it, and we choose to keep it, then it should not be `i32` but something like:
> 
> ```
> value: bool
> name: "should_truncate"
> description: >
>   If true perform truncation.  If false then round to `s` decimal places.  The default value is false.
> ```
> 
>  2. Should `s` be an option or a value argument?
> 
> In other words, is `round` a unary function with an option telling it how much to round to? Or is `round` a binary function?
> 
> My initial thought was "unary function" since that is how it is implemented in Acero. However, looking further, it appears that other engines have implemented this as a binary function. So I think binary function is the correct answer and this comment is no longer valid.
> 
> Let me know how you'd like to proceed on question 1.

@westonpace looking into the new options added (after addressing reviews), I think your comment on leaving functions from this definition would be better. Because the `truncate`  and `floor` are already defined as an option. 

But keeping `s` as optional, I am not so sure, we may need to include it, referring to [snowflake round function](https://docs.snowflake.com/en/sql-reference/functions/round.html).

File: extensions/functions_rounding.yaml [hunk 4]
```diff
@@ -27,3 +27,52 @@ scalar_functions:
           - value: fp64
             name: "x"
         return: fp64
+  -
+    name: "round"
+    description: >
+      Rounding the value `x` to s decimal places.
```
Comment (westonpace) [inline extensions/functions_rounding.yaml:33]:
```suggestion
      Rounding the value `x` to `s` decimal places.
```
This isn't markdown so I don't know if there is value in using backticks like this.  Do we do this elsewhere?  If we are using backticks consistently then `s` should also be quoted.

**Review by westonpace — CHANGES_REQUESTED** [review]
I agree that `s` needs to remain a second value argument given other engines work this way.  There are a few unique rounding options here that are not part of the IEEE rounding options.  These are `AWAY_FROM_ZERO`, `TIE_DOWN`, `TIE_UP`, `TIE_TOWARDS_ZERO`, `TIE_TO_ODD`.

Can we describe what these do somewhere?

**Comment by vibhatha** [comment]
> I agree that `s` needs to remain a second value argument given other engines work this way. There are a few unique rounding options here that are not part of the IEEE rounding options. These are `AWAY_FROM_ZERO`, `TIE_DOWN`, `TIE_UP`, `TIE_TOWARDS_ZERO`, `TIE_TO_ODD`.
> 
> Can we describe what these do somewhere?

Do you propose a documentation update or describe here itself in the function spec?

**Comment by westonpace** [comment]
> Do you propose a documentation update or describe here itself in the function spec?

In the function spec

**Comment by EpsilonPrime** [comment]
I like the description of the rounding options found in functions_arithmetic.yaml.  Perhaps we can utilize those here?

File: extensions/functions_rounding.yaml [hunk 5]
```diff
@@ -27,3 +27,92 @@ scalar_functions:
...
+                  halfway, tie away from zero.
+                - TRUNCATE: always round toward zero.
+                - CEILING: always round toward positive infinity.
+                - FLOOR: always round toward negative infinity.
+                - AWAY_FROM_ZERO: round negative values with FLOOR rule, round positive values with CEILING rule
+                - TIE_DOWN: round ties with FLOOR rule
+                - TIE_UP: round ties with CEILING rule
+                - TIE_TOWARDS_ZERO: round ties with TRUNCATE rule
+                - TIE_TO_ODD: round to nearest value; if exactly halfway, tie
+                  to the odd option.
```
Comment (vibhatha) [inline extensions/functions_rounding.yaml:72]: @westonpace @EpsilonPrime I added the spec here. And filled the missing fields. 

I copied the textual description 1:1 from the arithmetic PR. But I am not sure if we should state it 1:1 or describe differently. 

Maybe we need to update the arithmeitc rounding options too?

File: extensions/functions_rounding.yaml [hunk 6]
```diff
@@ -27,3 +27,92 @@ scalar_functions:
...
+              performed to the left side of the decimal point
+              as specified by `s`.
+        options:
+          rounding:
+            description: >
+              When a boundary is computed to lie somewhere between two values,
+              and this value cannot be exactly represented, this specifies how
+              to round it. For floating point numbers, it specifies the IEEE
+              754 rounding mode (as it does for all other floating point
+              operations). For integer types:
```
Comment (westonpace) [inline extensions/functions_rounding.yaml:58]:
```suggestion
              to round it:
```

Comment (westonpace) [reply extensions/functions_rounding.yaml:58]: These options are a superset of the IEEE 754 options.  Also, I'm not sure what you mean by "For integer types" since this function only applies to floating point types.  I think it's just best if we remove this part altogether.

**Review by westonpace — CHANGES_REQUESTED** [review]
Do we want implementations for integers?  It could be useful when `s` is negative.

File: extensions/functions_rounding.yaml [hunk 7]
```diff
@@ -27,3 +27,92 @@ scalar_functions:
           - value: fp64
             name: "x"
         return: fp64
+  -
+    name: "round"
+    description: >
+      Rounding the value `x` to `s` decimal places.
+    impls:
+      - args:
+          - value: fp32
```
Comment (EpsilonPrime) [inline extensions/functions_rounding.yaml:36]: Are there other pairs of argument types that we want to support?  I could see fp64 and i8 as potential options here.  With the ability to round anywhere in the number we may also want to include integer types for x.

**Comment by westonpace** [comment]
@vibhatha can you update this when you get a chance?

**Comment by vibhatha** [comment]
@westonpace will update it tomorrow. Sorry about the delay.

File: extensions/functions_rounding.yaml (hunk 7 above)
Comment (vibhatha) [reply extensions/functions_rounding.yaml:36]: Not entirely sure about `i8`, although with s being `i32` we have already added `fp32` and `fp64`. 
cc @westonpace

**Comment by vibhatha** [comment]
@westonpace sorry about the long wait. I updated the PR, although I see `branch is out-of-date with base branch`. I have rebased and did `--force` push.

File: extensions/functions_rounding.yaml (hunk 6 above)
Comment (westonpace) [reply extensions/functions_rounding.yaml:58]: You removed this test below (in the fp64/i32 kernel).  Can you also remove it here (in the fp32/i32 kernel)

File: extensions/functions_rounding.yaml (hunk 7 above)
Comment (westonpace) [reply extensions/functions_rounding.yaml:36]: For the moment let's assume it is safe to mandate that `s` be `i32`.  If we need to add additional kernels for `s` later then we can do so without much strife I hope.

However, we should allow `x` to be `i8, i16, i32, i64`.  So please add the additional 4 kernels.

**Review by westonpace — CHANGES_REQUESTED** [review]
Thanks for the update.  It still says out of date but a rebase should have fixed that.  Did you make sure you had updated your local `main` branch before rebasing?

**Comment by vibhatha** [comment]
> Thanks for the update. It still says out of date but a rebase should have fixed that. Did you make sure you had updated your local `main` branch before rebasing?

`main` and `round_func` branches are rebased with the `upstream/main`. Not sure what is going on here?

**Comment by vibhatha** [comment]
@westonpace I think now the branch is in sync with the base branch.

File: extensions/functions_rounding.yaml [hunk 8]
```diff
@@ -27,3 +27,248 @@ scalar_functions:
...
+            name: "x"
+            description: >
+              Numerical expression to be rounded.
+          - value: i32
+            name: "s"
+            description: >
+              Number of decimal places to be rounded to.
+
+              When `s` is a positive number, the rounding
+              is performed to a `s` number of decimal places.
```
Comment (westonpace) [inline extensions/functions_rounding.yaml:46]:
```suggestion
              When `s` is a positive number, nothing will happen
              since `x` is an integer value.
```

File: extensions/functions_rounding.yaml [hunk 9]
```diff
@@ -27,3 +27,248 @@ scalar_functions:
...
+
+              When `s` is a negative number, the rounding is
+              performed to the left side of the decimal point
+              as specified by `s`.
```
Comment (westonpace) [inline extensions/functions_rounding.yaml:50]:
```suggestion
              When `s` is a negative number, the rounding is
              performed to the nearest multiple of `10^(-s)`.
```

File: extensions/functions_rounding.yaml [hunk 10]
```diff
@@ -27,3 +27,248 @@ scalar_functions:
...
+                  halfway, tie away from zero.
+                - TRUNCATE: always round toward zero.
+                - CEILING: always round toward positive infinity.
+                - FLOOR: always round toward negative infinity.
+                - AWAY_FROM_ZERO: round negative values with FLOOR rule, round positive values with CEILING rule
+                - TIE_DOWN: round ties with FLOOR rule
+                - TIE_UP: round ties with CEILING rule
+                - TIE_TOWARDS_ZERO: round ties with TRUNCATE rule
+                - TIE_TO_ODD: round to nearest value; if exactly halfway, tie
+                  to the odd option.
```
Comment (westonpace) [inline extensions/functions_rounding.yaml:70]: What does `TIE_TO_EVEN` and `TIE_TO_ODD` mean for negative values of `s`?

CC @ianmcook

**Review by westonpace — CHANGES_REQUESTED** [review]
Thanks for adding integer kernels.  I think the description will need to be slightly different for these kernels.

File: extensions/functions_rounding.yaml (hunk 10 above)
Comment (vibhatha) [reply extensions/functions_rounding.yaml:70]: Ref: https://docs.oracle.com/cd/E57185_01/AGOHP/apbs03s18.html

Comment (vibhatha) [reply extensions/functions_rounding.yaml:70]: It is not quite clear from this definition either: https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html#function_round
:thinking:

Comment (westonpace) [reply extensions/functions_rounding.yaml:69]: I don't think we need to clarify.  I don't think we can easily do so without writing a novel.

**Review by westonpace — APPROVED** [review]
Alright, I think this looks good.  @EpsilonPrime did you want to take another look?

File: extensions/functions_rounding.yaml (hunk 10 above)
Comment (vibhatha) [reply extensions/functions_rounding.yaml:69]: I played with Mysql a bit, but with the round function, can we specify such options?

**Comment by vibhatha** [comment]
@westonpace a CI is failing and this happened a few times. Seems like unrelated, though  not very sure about it.

**Review by EpsilonPrime — COMMENTED** [review]
LGTM

**Comment by westonpace** [comment]
> @westonpace a CI is failing and this happened a few times. Seems like unrelated, though not very sure about it.

Unrelated to this PR.  I created #439 which should address it.
